### ISSUE TITLE
Unable to get webpack.NormalModuleReplacementPlugin working when building for electron

### ISSUE THREAD

**Issue by bdiz** [nklayman/vue-cli-plugin-electron-builder#31]
**Describe the bug**
I am using a single code base to target web and electron. I have a need to import different modules depending on the build target. I am attempting to use webpack.[NormalModuleReplacementPlugin](https://webpack.js.org/plugins/normal-module-replacement-plugin/) to meet this need as it will not include code in the bundle which is not needed for build targets which it is not currently building.

**To Reproduce**
Create a vue.config like this:
```javascript
var webpack = require('webpack');

module.exports = { 
  configureWebpack: () => {

    let plugins = [ 

      new webpack.NormalModuleReplacementPlugin(/(.*)-VUE_APP_PLATFORM(\.*)/, function(resource) {
        resource.request = resource.request.replace(/-VUE_APP_PLATFORM/, `-${process.env.VUE_APP_PLATFORM}`);
      })  
    ];  

    return {
      plugins,
    };  

  }
};
```
Create a new env file called .env.electron-production:
```
NODE_ENV=production
VUE_APP_PLATFORM=electron
```
Optionally, create .env.web-production if you want to reproduce the pass case.
```
NODE_ENV=production
VUE_APP_PLATFORM=web
```
Create two files such as, db-web.js and db-electron.js. 

Import these modules in your project code using the line:
```javascript
import db from "../db-VUE_APP_PLATFORM";
```
Run vue-cli-plugin-electron-builder: `vue-cli-service build:electron --mode electron-production`. (I've also tried first exporting `VUE_APP_PLATFORM=electron`, but it does not effect the result).

This error is encountered:
```
  ERROR in ./src/odm.js
  Module not found: Error: Can't resolve '../db-VUE_APP_PLATFORM' in 'src'
```

**Expected behavior**
Expectation was the module would be found without error. 

When building for web with `vue-cli-service build --mode web-production`, this plugin succeeds and the modules are found by webpack.

**Environment:**
 - OS and version: Ubuntu 16.10
 - Node version: v10.5.0
 - npm version: 6.1.0
 - yarn version (if used): 
 - vue-cli-plugin-electron-builder version : 0.3.2
 - vue/cli-service: 3.0.0-rc.3
 - other vue plugins used: none
 - custom config for vcp-electron-builder: none
 - (if possible) link to your repo: not possible

Thanks for this plugin!

**Comment by nklayman** [comment]
This won't work because VCP-electron-builder uses electron-webpack. Electron-webpack generates it's own config, independent of that for your regular app. You can use the [webpackConfig property](https://github.com/nklayman/vue-cli-plugin-electron-builder/tree/legacy#electron-webpack) under VCP-electron-builders plugin options. A better alternative is to use the 1.0.0-beta of VCP electron-builder. It doesn't use electron-webpack, instead modifying your regular config. If you use the 1.0.0 beta, this issue should be automatically solved. It is just as stable as 0.3.2, but is much better in every way. Let me know if either solution works.

**Comment by bdiz** [comment]
vue-cli-service build:electron is looking better. Thanks!

vue-cli-service serve:electron I get this error:
```
 ERROR  Failed to compile with 3 errors                                                                                           12:37:14 PM

These relative modules were not found:

* ../db-VUE_APP_PLATFORM in ./src/odm.js
...
```
Also, can you explain your comment about `--mode`? In beta 4 will different .env files work? What is the purpose of defining IS_WEBPACK? I may not understand because I don't know what is happening under the hood.

**Comment by nklayman** [comment]
Read the [How It Works](https://github.com/nklayman/vue-cli-plugin-electron-builder/tree/master#how-it-works) section of the docs to learn what happens under the hood. Since it triggers the serve command, not all arguments are passed though. I can allow the `--mode` argument to get passed through to build/serve commands if you would like. The `IS_WEBPACK` variable would be replaced with `true` during webpack bundling of electron builds only.

**Comment by bdiz** [comment]
Yes, I'd like to get my .env files to vue-cli-service build and serve. Adding `--mode` pass through would be great. Do you think this is why I'm getting the error with vue-cli-service serve:electron?

**Comment by nklayman** [comment]
Try `window.process.env` for the env varaibles.

**Comment by bdiz** [comment]
And I got a initial route that looks odd to me (due to the `app.asar/dist_electron') when the electron app opens: `/home/bdiz/my-app/dist_electron/linux-unpacked/resources/app.asar/dist_electron/bundled/index.html`

**Comment by nklayman** [comment]
The reason it is set to `.../dist_electron/bundled` is because all files that go through webpack (including files in the `public` directory) end up in `dist_electron/bundled` as not to corrupt the `dist` directory with non SPA compatible code. Electron builder has this nested in the `app.asar` because it needs to access your `node_modules` and `package.json` which are not nested, so the root directory is set to `./`. I will look into copying these into `dist_electron/bundled` so that the files end up in the root directory of the `app.asar`.

**Comment by bdiz** [comment]
Hi @nklayman , nice docs.

I tried out beta.4 but I am still getting the same failing signature with `serve:electron`.

**Comment by nklayman** [comment]
I realized that the mode is set through the plugin service, not the command itself. This means that all the changes I made didn't really do anything. The good news is I found the solution to your problem, just change your `vue.config.js` to look like this:

```javascript
var webpack = require('webpack')

module.exports = {
  configureWebpack: () => {
    let plugins = [
      new webpack.NormalModuleReplacementPlugin(
        /(.*)-VUE_APP_PLATFORM(\.*)/,
        function(resource) {
          resource.request = resource.request.replace(
            /-VUE_APP_PLATFORM/,
            `-${process.env.VUE_APP_PLATFORM}`
          )
        }
      )
    ]

    return {
      plugins
    }
  },
  pluginOptions: {
    electronBuilder: {
      chainWebpackRendererProcess(config) {
        config.plugin('modules').use(webpack.NormalModuleReplacementPlugin, [
          /(.*)-VUE_APP_PLATFORM(\.*)/,
          function(resource) {
            resource.request = resource.request.replace(
              /-VUE_APP_PLATFORM/,
              `-${process.env.VUE_APP_PLATFORM}`
            )
          }
        ])
        return config
      }
    }
  }
}

```

**Comment by bdiz** [comment]
I pasted that code and it did remove that error signature. However, I have other configs and module rules that are now producing errors so it looks like I need to duplicate those configs into `pluginOptions.electronBuilder` as well. Is there any way to avoid this duplication? ie somehow pass my existing configs and rules to electron builder?

**Comment by nklayman** [comment]
The existing rules should be transferred to electron as it just calls the serve command with a few modifications. I have no idea why this isn't working, try modifying your config through chaining and see if that helps.

**Comment by bdiz** [comment]
Your guess was right. Chaining was the answer to get serve working without duplicating configs.

vue.config.js:
```javascript
var webpack = require('webpack');

module.exports = { 
  chainWebpack: (config) => {

    config.plugin('modules').use(webpack.NormalModuleReplacementPlugin, [
      /(.*)-VUE_APP_PLATFORM(\.*)/,
      function(resource) {
        resource.request = resource.request.replace(
          /-VUE_APP_PLATFORM/,
          `-${process.env.VUE_APP_PLATFORM}`
        )
      }
    ]);

  }
};
```

**Comment by bdiz** [comment]
@nklayman In regards to https://github.com/nklayman/vue-cli-plugin-electron-builder/issues/31#issuecomment-403141118 would you like me to add an issue to track addition of `vue-cli-service build` pass through options? Most importantly the `--mode` option.

**Comment by bdiz** [comment]
In beta.6 I'm getting an error with `vue-cli-service build:electron --mode my-mode`:

```
Building app with electron-builder:

  • electron-builder version=20.28.1

/home/user/my-app/node_modules/app-builder-lib/src/index.ts:42
      throw new InvalidConfigurationError(`Unknown option "${optionName}"`)
            ^
Error: Unknown option "mode"
```

**Comment by nklayman** [comment]
Fixed in 79a1be9b04f5adf442b233dc1486eeebb9179275. Sorry about that, didn't realize you were getting that error. I am releasing `1.0.0-beta.7` now.

**Comment by bdiz** [comment]
Thank you. `--mode` is getting passed through. 

Next switch I'm using and is not getting passed through is `--modern`.

`Error: Unknown option "modern"`

**Comment by nklayman** [comment]
`--modern` is enabled by default

**Comment by bdiz** [comment]
Ok. Cool.
